feat: force snapshot under memory pressure

praveen-influx opened the pull request:
- The core of the change is to introduce another method `force_flush_buffer` in `Wal` trait. This gives a handle to choose when to kick off snapshot.
- A higher level background loop is introduced that checks the overall table buffer size every `N` seconds and if it is greater than a threshold (`X`) then it calls `force_flush_buffer`. Both `N` and `X` are configurable through cli. `N` defaults to 10s and `X` defaults to 70%
- Some refactoring of the code went on to make sure the calls made via `Wal` trait to flush buffer and cleanup any snapshot is reused across both branches (forcing snapshot and normal wal buffer flush)

closes: https://github.com/influxdata/influxdb/issues/25685

pauldix commented on influxdb3/src/commands/serve.rs:
```diff
@@ -293,6 +296,25 @@ pub struct Config {
+    /// Interval to check buffer size (and compare with `force_snapshot_mem_threshold`)
+    #[clap(
+        long = "force-snapshot-interval",
+        env = "INFLUXDB3_FORCE_SNAPSHOT_INTERVAL",
```
I don't think this needs to be a configuration option. Just have it as a constant. You can then either initialize with that or, in tests, with something smaller.

praveen-influx replied:
I've removed this interval in the other [PR](https://github.com/influxdata/influxdb/pull/25727), although I was planning to use it in e2e tests.

pauldix commented on influxdb3_wal/src/lib.rs:
```diff
@@ -809,6 +856,10 @@ impl WalContents {
     }
+
+    pub fn is_default(&self) -> bool {
+        self.min_timestamp_ns == i64::MAX && self.max_timestamp_ns == i64::MIN
```
default for Gen1 is 10 minutes, so it should be a check that max_ns - min_ns = 600000000000

pauldix commented on influxdb3_wal/src/object_store.rs:
```diff
@@ -450,21 +466,24 @@ impl FlushBuffer {
-        });
-
-        let snapshot = match self.snapshot_tracker.snapshot() {
+        if !force_snapshot {
```
Why do we only add the WAL period if we're not forcing the snapshot? I think a comment here would be useful

pauldix commented on influxdb3_wal/src/snapshot_tracker.rs:
```diff
@@ -118,6 +123,29 @@ impl SnapshotTracker {
         })
     }
 
+    fn snapshot_up_to_last_wal_period(&mut self) -> Option<SnapshotInfo> {
```
This function name is a bit off. During normal snapshot operation, we want to snapshot the flush size number of wal periods. But we want to leave behind flush size / 2 periods. So that means we wait until we have flush size + flush size / 2 periods and then we take the oldest flush size periods, leaving behind flush size / 2.

If we're in a situation where we can't flush the WAL and leave behind periods because the time stamps of the data are all interleaved, we flush everything, except the most recent wal period. We do this because the buffer snapshots what is in it and then puts the most recent period into the snapshot.

In the case of forcing a snapshot, we don't need to check `should_snapshot`, we just treat it like the situation where we have 3x the flush size. So we want to snapshot everything minus the last wal period.

praveen-influx replied:
Thanks for the explanation - I've addressed it [here](https://github.com/influxdata/influxdb/blob/3f557b816780fcdf11f47113c78021ff465af017/influxdb3_wal/src/snapshot_tracker.rs#L60-L91) in the other [PR](https://github.com/influxdata/influxdb/pull/25727)

pauldix reviewed (commented):
Left a comment about the logic, I think there's one bit to be cleaned up. Will have a quick call to walk through.

praveen-influx commented:
Thanks @pauldix - I'll close this in favour of https://github.com/influxdata/influxdb/pull/25727, it takes a different approach so no point in continuing with the same PR. I've also addressed your comments that are still valid in the other PR.
